api/src/main/java/io/druid/data/input/impl/prefetch/PrefetchableTextFilesFirehoseFactory.java

```diff
@@ -0,0 +1,248 @@
+ * future reads.
+ * <br/>
+ * - Fetching: when it reads all cached data, it fetches remaining objects into a local disk and reads data from
+ * them.  For the performance reason, prefetch technique is used, that is, when the size of remaining fetched data is
+ * smaller than {@link #prefetchTriggerBytes}, a background prefetch thread automatically starts to fetch remaining
+ * objects.
+ * <br/>
+ * - Retry: if an exception occurs while downloading an object, it retries again up to {@link #maxFetchRetry}.
+ * <p/>
+ *
+ * This implementation can be useful when the cost for reading input objects is large as reading from AWS S3 because
+ * batch tasks like IndexTask or HadoopIndexTask can read the whole data twice for determining partition specs and
+ * generating segments if the intervals of GranularitySpec is not specified.
+ * <br/>
+ * Prefetching can be turned on/off by setting maxFetchCapacityBytes.  Depending on prefetching is enabled or
+ * disabled, the behavior of the firehose is different like below.
+ * <p/>
+ *
+ * 1. If prefetch is enabled, this firehose can fetch input objects in background.
+ * <br/>
+ * 2. When next() is called, it first checks that there are already fetched files in local storage.
+ * <br/>
+ *   2.1 If exists, it simply chooses a fetched file and returns a {@link LineIterator} reading that file.
+ *   <br/>
+ *   2.2 If there is no fetched files in local storage but some objects are still remained to be read, the firehose
+ *   fetches one of input objects in background immediately. If an IOException occurs while downloading the object,
+ *   it retries up to the maximum retry count. Finally, the firehose returns a {@link LineIterator} only when the
+ *   download operation is successfully finished.
+ *   <br/>
+ * 3. If prefetch is disabled, the firehose returns a {@link LineIterator} which directly reads the stream opened by
+ * {@link #openObjectStream}. If there is an IOException, it will throw it and the read will fail.
+ */
+public abstract class PrefetchableTextFilesFirehoseFactory<ObjectType>
+    extends AbstractTextFilesFirehoseFactory<ObjectType>
+{
+  private static final Logger LOG = new Logger(PrefetchableTextFilesFirehoseFactory.class);
+  private static final long DEFAULT_MAX_CACHE_CAPACITY_BYTES = 1024 * 1024 * 1024; // 1GB
+  private static final long DEFAULT_MAX_FETCH_CAPACITY_BYTES = 1024 * 1024 * 1024; // 1GB
+  private static final long DEFAULT_FETCH_TIMEOUT = 60_000; // 60 secs
+  private static final int DEFAULT_MAX_FETCH_RETRY = 3;
+
+  private final CacheManager<ObjectType> cacheManager;
+  private final long maxFetchCapacityBytes;
+  private final long prefetchTriggerBytes;
+  private final long fetchTimeout;
+  private final int maxFetchRetry;
+
+  private List<ObjectType> objects;
+
+  private static ExecutorService createFetchExecutor()
+  {
+    return Executors.newSingleThreadExecutor(
```

Comment:
`Execs` may have existing shortcut for this

Reply:
The druid-api module doesn't have a dependency on druid-common.

Reply:
Druid-api shouldn't actually exist anymore, since PublicApi and ExtensionPoint annotations are added.

Reply:
I mean, adding a dependency to druid-api can increase the size of user binaries which use druid-api. ```Execs``` is currently in druid-common, and its size is not large so adding to druid-api may be fine. However, I still think it's better to move ```Execs``` to java-util rather than adding a dependency. What do you think?

Reply:
I think it makes sense to either move `Execs` to java-util, or skip the shortcut for now.

I also think that what @leventov is saying is that in the future, we probably want to collapse java-util, druid-api, and druid-common into a single module anyway.

Reply:
Fwiw: the history is that java-util is separate because it used to be an independent project, but now it's part of Druid. And druid-api is separate because it was meant to be the module that held all extension points, but this didn't really work out, since it turns out we have extension points in other modules too.

Reply:
Thanks @gianm. I thought druid-common is a library for internal use in druid. If it is also intended for public use, I agree on collapsing all three modules.

Reply:
It's intended for use by extensions, really, along with any other module that has any `@PublicApi` or `@ExtensionPoint` annotations in it. I guess that counts as public.

Reply:
Thanks. Moved Execs to java-util and changed to use it.